### PR TITLE
Feature/exposing quantiles

### PR DESCRIPTION
**Pull request by warebot**
Hey guys.
I started working on this "issue" because it is requirement that we currently have.(https://github.com/prometheus/client_java/issues/85)

No tests were broken, new tests were added and the outcome seems to be accurate.
The solutions were inspired by other open source implementations/ideas (I'm pretty sure I need to cite them in the code).

Would highly appreciate your feedback re:what can be improved / needs to be changed.
Please pardon my lack of Github etiquette.
 
Mucho gracias!

### CODE REVIEW

File: simpleclient/src/main/java/io/prometheus/client/Reservoir.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package io.prometheus.client;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Created by warebot on 12/21/15.
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Reservoir.java]: This comment doesn't belong here, the AUTHORs file and source control history cover this.

File: simpleclient/src/main/java/io/prometheus/client/Reservoir.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
+package io.prometheus.client;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Reservoir.java]: This is a Java 7 feature, this library is targeted at Java 6 as 2/3 of our users still use Java 6. You'll need to find a way to do this within the Java 6 featureset.

File: simpleclient/src/main/java/io/prometheus/client/Summary.java [hunk 3]
```diff
@@ -1,18 +1,16 @@
 package io.prometheus.client;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Summary.java]: Please list all the imports explicitly so it's easy to figure out where they're coming from

File: simpleclient/src/main/java/io/prometheus/client/Summary.java [hunk 4]
```diff
@@ -1,18 +1,16 @@
...
  * <p>
  * Example of uses for Summaries include:
  * <ul>
- *  <li>Response latency</li>
- *  <li>Request size</li>
+ * <li>Response latency</li>
+ * <li>Request size</li>
  * </ul>
- * 
+ * <p>
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Summary.java]: Extra `<p>`

File: simpleclient/src/main/java/io/prometheus/client/Reservoir.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    public double getQuantile(double quantile, double[] data) {
+        if (quantile < 0.0 || quantile > 1.0 || Double.isNaN(quantile)) {
+            throw new IllegalArgumentException(quantile + " is not in [0..1]");
+        }
+
+        int size = data.length;
+
+        if (size == 0) {
+            return 0.0;
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Reservoir.java]: NaN is a better answer here, as we've no data. This is what the other clients do.

File: simpleclient/src/main/java/io/prometheus/client/Summary.java [hunk 6]
```diff
@@ -38,133 +36,208 @@
...
+
+            // If the quantiles is a desired calculation then add the quantile calculations to the Summary.
+            if (this.shouldCalculateQuantiles()) {
+                // Since `labelNames` is declared final, we need to create a new list that includes our "quantile"
+                // label.
+                double[] data = c.getValue().reservoir.getSnapshot();
+                Arrays.sort(data);
+
+                List<String> labelNamesWithQuantile = new ArrayList<String>(labelNames);
+                labelNamesWithQuantile.add("quantile");
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/Summary.java]: The quantile label needs to be banned up in Builder, see how it's done in Histogram.

**Comment by brian-brazil** [comment]
Thanks for the PR. This implementation randomly replaces previous samples in the reservoir. The Go implementation (https://github.com/prometheus/client_golang/blob/master/prometheus/summary.go) expires samples over time, and I think we should do that here for both consistency and so that the quantile is reasonably recent.

@beorn7

**Comment by openendings** [comment]
@warebot Where can the new pull request be found?

**Comment by openendings** [comment]
I'd be interested, yeah; thank you!

**Pull request closed without merging**